smart_importer/hooks.py

```diff
@@ -36,18 +36,17 @@ def apply_hooks(importer, hooks):
     unpatched_extract = importer.extract
 
     @wraps(unpatched_extract)
-    def patched_extract_method(file, existing_entries=None):
+    def patched_extract_method(filepath, existing=None):
         logger.debug("Calling the importer's extract method.")
-        imported_entries = unpatched_extract(
-            file, existing_entries=existing_entries
-        )
+        imported_entries = unpatched_extract(filepath, existing=existing)
 
         for hook in hooks:
             imported_entries = hook(
-                importer, file, imported_entries, existing_entries
+                importer, filepath, imported_entries, existing
             )
 
         return imported_entries
 
     importer.extract = patched_extract_method
+    importer.deduplicate = lambda entries, existing: None
```

Comment:
Why? I think this line should be removed, we shouldn't silently remove this method

Reply:
Maybe there is a better way to do it. The problem is that the DuplicateDetector hook from smart_importer doesn't work with this, because that one overrides it. All my duplicatedetector logic didn't work anymore because of this.

Reply:
Maybe the hook should allow to override both the extract and the deduplicate method and the DuplicateDetector should override the deduplicate?

Reply:
I'll wrap this in an if-block with a warning if DuplicateDetector is used then. With the deduplicate method on the importer itself, it should be pretty simple to now remove any use of DuplicateDetector.

Reply:
What I liked with the hooks was that I could reuse (custom) duplicate detection logic across many different importers. Maybe I'll need to create my own wrapper instead of the hooks to allow reusing deduplication logic across importers.

Reply:
I liked them as well. But I think it's also hacky to patch these methods and we'd now have three kinds of "hooks": 1) "hooks" as beangulp and Fava use them - applied across all importers, 2) "hooks" that should patch the extract method of a single importer, and 3) "hooks" that should also patch/remove the deduplicate method (our DuplicateDetector). I think we should avoid introducing 3) and probably switch to some alternative of 2) as a recommended way to use the package overall.

Could you maybe create a shared base class for your importers that contains the deduplication logic you want? Or just patch the deduplicate method directly?

Reply:
Actually now looking at it once more, moving this code into the importers actually makes sense, so I'll move the deduplication logic into a common part in my importers and then can reuse it across importers. I'm wondering if should even keep the deduplication hook.

Reply:
>  I'm wondering if should even keep the deduplication hook.

I'd be for removing it - but I don't think we need to do / should to do that right now to make upgrading easier. I'll add a @deprecated decorator to it.

Reply:
Now it works fine in plain beangulp but doesn't work correctly with fava, I think the issue is that the beangulp implementation is now setting the __duplicate__ field to the original of the duplicated transaction and fava expects to receive True instead.

Reply:
Created https://github.com/beancount/fava/pull/1937 but doesn't pass lint yet, would be great if you can help there. It works fine in my tests, but there is probably a more correct way to write it so it actually passes linting.